### PR TITLE
Speedy: make the compiler/engine ready for Nat types

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR advances the Numeric implementation #2289.

With this PR the engine is able to handle the `Nat` type as normal data, enabling hence the possibility to handle dynamically scale for Numerics. 

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [X] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala [hunk 1]
```diff
@@ -38,7 +38,7 @@ object SExpr {
     * that is, SEVar(1) points to the top-most value in the environment.
     * https://en.wikipedia.org/wiki/De_Bruijn_index
     */
-  final case class SEVar(index: Int) extends SExpr {
+  final case class SVar(index: Int) extends SExpr {
```
Comment (hurryabit) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala]: I'm generally not a big fan of doing a renaming like this and actual changes in the same PR. I think I understand why you want to rename it but it might be better to the in separate PR before the one with the actual changes.

Comment (dajmaki) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala]: I don't like this at all. It conflicts with the naming convention used for SValue. Please change it back.

Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala]: I created #2657 to address this comment. 
I will switch the PR to ready to review once #2657 is merged.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala [hunk 2]
```diff
@@ -174,6 +181,8 @@ object SValue {
 
   final case class SMap(value: HashMap[String, SValue]) extends SValue
 
+  final case class STNat(n: Int) extends SValue
```
Comment (dajmaki) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala]: add a comment explaining what this is for and how and where it is used.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 3]
```diff
@@ -855,13 +880,13 @@ final case class Compiler(packages: PackageId PartialFunction Package) {
     *
     * For example:
     *   SELet(...) in
-    *     SEAbs(2, SEVar(4))
+    *     SEAbs(2, SVar(4))
     * =>
     *   SELet(...) in
     *     SEMakeClo(
-    *       Array(SEVar(2)), (capture 2nd value)
+    *       Array(SVar(2)), (capture 2nd value)
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala]: fixed

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 4]
```diff
@@ -1003,7 +1028,7 @@ final case class Compiler(packages: PackageId PartialFunction Package) {
       expr match {
         case SEVar(i) =>
           if (i < 1 || i > bound) {
-            throw CompileError(s"validate: SEVar: index $i out of bound $bound")
+            throw CompileError(s"validate: SVar: index $i out of bound $bound")
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala]: fixed

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala]: done
